library/std/src/sys/uefi/stdio.rs

```diff
@@ -4,45 +4,48 @@ use crate::mem::MaybeUninit;
 use crate::os::uefi;
 use crate::ptr::NonNull;
 
-const MAX_BUFFER_SIZE: usize = 8192;
+pub struct Stdin {
+    pending: Option<char>,
+}
 
-pub struct Stdin;
 pub struct Stdout;
 pub struct Stderr;
 
 impl Stdin {
     pub const fn new() -> Stdin {
-        Stdin
+        Stdin { pending: None }
     }
 }
 
 impl io::Read for Stdin {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
         let st: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table().cast();
         let stdin = unsafe { (*st.as_ptr()).con_in };
 
-        // Try reading any pending data
-        let inp = match read_key_stroke(stdin) {
-            Ok(x) => x,
-            Err(e) if e == r_efi::efi::Status::NOT_READY => {
-                // Wait for keypress for new data
-                wait_stdin(stdin)?;
-                read_key_stroke(stdin).map_err(|x| io::Error::from_raw_os_error(x.as_usize()))?
+        // Write any pending character
+        if let Some(ch) = self.pending {
+            if ch.len_utf8() > buf.len() {
+                return Ok(0);
```

Comment:
You could return UTF-8 data byte-by-byte, rather than refusing to return it if it does not fit into the buffer. Any reason not to do this?

The thing is, I am not aware of any implementation that refuses to return data bytewise. This can easily break streaming implementations that fetch into their own ring-buffer and just happen to have less than 4-bytes of space.

Reply:
Btw., I am ok with your implementation. I am just saying I think we can do better!

Reply:
Rust implementations generally avoid emitting malformed UTF-8, as far as I have seen, as this would cause the entire buffer to now fail `str::from_utf8`.

Reply:
This means that fetching it bytewise makes it library UB to use `str::from_utf8_unchecked`, and a lot of Rust code relies on the "data that has made it through Rust's string validation once, remains valid UTF-8" property. 4 bytes simply doesn't seem worth the harm to soundness and reasoning.

Reply:
The problem that does require this _anyways_, despite my grievances, is due to

> If the return value of this method is [Ok(n)](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Ok), then implementations must guarantee that 0 <= n <= buf.len(). A nonzero n value indicates that the buffer buf has been filled in with n bytes of data from this source. If n is 0, then it can indicate one of two scenarios:
> 
> - This reader has reached its “end of file” and will likely no longer be able to produce bytes. Note that this does not mean that the reader will always no longer be able to produce bytes. As an example, on Linux, this method will call the recv syscall for a [TcpStream](https://doc.rust-lang.org/std/net/struct.TcpStream.html), where returning zero indicates the connection was shut down correctly. While for [File](https://doc.rust-lang.org/std/fs/struct.File.html), it is possible to reach the end of file and get zero as result, but if more data is appended to the file, future calls to read will return more data.
> - The buffer specified was 0 bytes in length.

For some ghastly reason we chose `Ok(0)` to mean "either EOF or 0 bytes were possible to write, who knows!" So, implementations of `Read::read`, lacking another specified behavior, can only return `Ok(0)` to signal EOF.

Reply:
Because `stdin` is a byte-stream, not a text-stream. And it is used as such very commonly to feed non-text data to an application.

Reply:
After looking into internals, the methods like `read_to_string` seem to do their own utf8 validation on stdin (https://github.com/rust-lang/rust/blob/fcca978aa0c9a7443649305dd0ed48c4691a1e89/library/std/src/io/mod.rs#L382). So maybe it is fine to return illformated data? I don't think read or any upper levels in Rust stdio abstraction assume that this data is valid UTF-8 (unless you read into a string, in which case the necessary checks are performed).

The reason for using char in this implementation in the first place is to perform utf16 character to utf8 character conversion. Not to ensure any kind of utf8 validity.

Reply:
Yeah, I agree.

Reply:
Indeed, let's go with that.